### PR TITLE
fix(excel2resources, excel2properties): make more robust, correct schemas

### PR DESCRIPTION
**Pull request by jnussbaum**
resolves DEV-1040

### CODE REVIEW

File: MANIFEST.in [hunk 1]
```diff
@@ -4,4 +4,3 @@ include knora/dsplib/schemas/lists-only.json
 include knora/dsplib/schemas/resources-only.json
 include knora/dsplib/schemas/properties-only.json
 include knora/dsplib/schemas/data.xsd
-include knora/dsplib/utils/language-codes-3b2_csv.csv
```
Comment (jnussbaum) [inline MANIFEST.in:7]: I removed that file in an earlier PR, and forgot to remove it also here.

File: docs/dsp-tools-excel.md [hunk 2]
```diff
@@ -58,15 +58,16 @@ The Excel sheet must have the following structure:
 
 The expected columns are:
 
-- `name` : The name of the property
-- `super` : The base property of the property
-- `object` : If the property is derived from `hasValue`, the type of the property must be further specified by the 
+- `name`: The name of the property
+- `super`: The base property/ies of the property, separated by commas
```
Comment (jnussbaum) [inline docs/dsp-tools-excel.md:62]: multiple superclasses must be allowed

File: docs/dsp-tools-excel.md [hunk 3]
```diff
@@ -58,15 +58,16 @@ The Excel sheet must have the following structure:
...
 object it takes, e.g. `TextValue`, `ListValue`, or `IntValue`. If the property is derived from `hasLinkTo`, 
 the `object` specifies the resource class that this property refers to.
-- `en`, `de`, `fr`, `it` : The labels of the property in different languages, at least one language has to be provided
+- `en`, `de`, `fr`, `it`: The labels of the property in different languages, at least one language has to be provided
 - `comment_en`, `comment_de`, `comment_fr`, `comment_it`: optional comments in the respective language 
-- `gui_element` : The GUI element for the property
-- `hlist` : In case of list values: the name of the list
+- `gui_element`: The GUI element for the property
+- `gui_attributes` (optional): The gui_attributes in the form "attr: value, attr: value". In case of ListValues, the 
+  name of the list can be given as "hlist: listname" (according to the pattern), or simply as "listname".
```
Comment (jnussbaum) [inline docs/dsp-tools-excel.md:70]: For backwards compatibility, I renamed the last column from 'hlist' to 'gui_attributes', and allowed the simple 'listname' instead of 'hlist: listname'. Like this, this PR doesn't break old excel files.
It is necessary to support gui_attributes, because they are required in some cases.

File: knora/dsplib/schemas/lists-only.json [hunk 4]
```diff
@@ -1,5 +1,5 @@
 {
-    "$schema": "https://json-schema.org/draft-07/schema",
+    "$schema": "http://json-schema.org/draft-07/schema#",
```
Comment (jnussbaum) [inline knora/dsplib/schemas/lists-only.json:2]: With `https`, the file is not found, which produces the following warning:
`DeprecationWarning: The metaschema specified by $schema was not found. Using the latest draft to validate, but this will raise an error in the future.`

File: knora/dsplib/schemas/properties-only.json [hunk 5]
```diff
@@ -82,6 +82,7 @@
                                 "ListValue",
                                 "Region",
                                 "Resource",
+                                "Representation",
```
Comment (jnussbaum) [inline knora/dsplib/schemas/properties-only.json:85]: In the main schema, `Representation` had been added earlier, but it was forgotten here.

File: knora/dsplib/utils/excel_to_json_properties.py [hunk 6]
```diff
@@ -44,7 +45,8 @@ def properties_excel2json(excelfile: str, outfile: str) -> list[dict[str, Any]]:
     # load file
     wb = load_workbook(filename=excelfile, read_only=True)
     sheet = wb.worksheets[0]
-    props = [row_to_prop(row) for row in sheet.iter_rows(min_row=2, values_only=True, max_col=13)]
+    props = [row_to_prop(row) for row in sheet.iter_rows(min_row=2, values_only=True, max_col=13)
+             if any(row) and any([re.search(r'[A-Za-z]+', elem) for elem in row if isinstance(elem, str)])]
```
Comment (jnussbaum) [inline knora/dsplib/utils/excel_to_json_properties.py:49]: Without this check, openpyxl goes through the entire Excel sheet, which produces a lot of rows that only contain `None` or cells with whitespaces in it. The regex is to check if there are actually words in a cell, not only whitespaces.

File: .github/workflows/check-pr-title.yml [hunk 7]
```diff
@@ -12,7 +12,7 @@ jobs:
       # check PR title
       - uses: deepakputhraya/action-pr-title@master
         with:
-          regex: '([a-z])+(\(([a-z\-_ ])+\))?!?: [a-z]([a-zA-Z-\.\d \(\)\[\]#_,])+$' # Regex the title should match.
+          regex: '([a-z])+(\(([0-9a-z\-_, ])+\))?!?: [a-z]([a-zA-Z-\.\d \(\)\[\]#_,])+$' # Regex the title should match.
```
Comment (BalduinLandolt) [inline .github/workflows/check-pr-title.yml:15]: should comma and space be allowed in the scope of a PR title? and do we still need the PR title workflow? (at times I'd wish we would have it in all repos... but having it only in this one seems a bit weird to me)

File: knora/dsplib/utils/excel_to_json_properties.py (hunk 6 above)
Comment (BalduinLandolt) [reply knora/dsplib/utils/excel_to_json_properties.py:49]: what about numbers etc.? I think there should be a method on strings to check if it's whitespace. This might prove more reliable in some edge cases (and may even be faster)

File: knora/dsplib/utils/excel_to_json_properties.py [hunk 8]
```diff
@@ -70,35 +72,51 @@ def row_to_prop(row: tuple[str, str, str, str, str, str, str, str, str, str, str
     Returns:
         prop (JSON): the property in JSON format
     """
-    name, super_, object_, en, de, fr, it, comment_en, comment_de, comment_fr, comment_it, gui_element, hlist = row
+    name, super_, object_, en, de, fr, it, comment_en, comment_de, comment_fr, comment_it, gui_element, gui_attributes = row
     labels = {}
     if en:
-        labels['en'] = en
+        labels['en'] = en.strip()
```
Comment (BalduinLandolt) [inline knora/dsplib/utils/excel_to_json_properties.py:78]: you do lots of `.strip()` all over the place, so you might consider just striping each string when you read the file.

File: knora/dsplib/utils/excel_to_json_resources.py [hunk 9]
```diff
@@ -48,7 +48,8 @@ def resources_excel2json(excelfile: str, outfile: str) -> None:
 
     # get overview
     sheet = wb['classes']
-    resource_list = [c for c in sheet.iter_rows(min_row=2, values_only=True)]
+    resource_list = [c for c in sheet.iter_rows(min_row=2, values_only=True)
+                     if any(c) and any([re.search(r'[A-Za-z]+', elem) for elem in c if isinstance(elem, str)])]
```
Comment (BalduinLandolt) [inline knora/dsplib/utils/excel_to_json_resources.py:52]: as above: if you just strip each string here, you can just check if it's non-empty, and you don't run the risk of a edge case where the regex misses a valid string

File: test/unittests/test_excel_to_properties.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            "hasAnthroponym  ",
+            " hasValue, dcterms:creator   ",
+            " TextValue ",
+            "anthroponym",
+            "",
+            "Anthroponyme",
+            "",
+            " A strange chance put me in possession of this journal.  ",
+            "",
+            "",
```
Comment (BalduinLandolt) [inline test/unittests/test_excel_to_properties.py:35]: adding a number only/umlaut only field here would probably get lost because of the regex only looking as a-zA-Z

**Review by BalduinLandolt — APPROVED** [review]
Looks good overall. I'm concerned the regex might miss some edge cases... and stripping all strings in the beginning would spare you from doing so all over the place.

File: knora/dsplib/utils/excel_to_json_properties.py (hunk 6 above)
Comment (jnussbaum) [reply knora/dsplib/utils/excel_to_json_properties.py:49]: It's not necessary to check edge cases, because as soon as only one letter is present in one cell, the entire row is processed. Actually, this is rather too permissive than too strict.

File: .github/workflows/check-pr-title.yml (hunk 7 above)
Comment (jnussbaum) [reply .github/workflows/check-pr-title.yml:15]: In the present case, it's necessary. (Otherwise I wouldn't have done it :-) 
I find it good to have the PR title workflow, so we get clean and consistent commit messages in the main branch. Surely, other repos would profit from such a workflow!

File: knora/dsplib/utils/excel_to_json_properties.py (hunk 8 above)
Comment (jnussbaum) [reply knora/dsplib/utils/excel_to_json_properties.py:78]: That sounds like a good idea, but I don't see an easy solution to do that.

File: knora/dsplib/utils/excel_to_json_resources.py (hunk 9 above)
Comment (jnussbaum) [reply knora/dsplib/utils/excel_to_json_resources.py:52]: It's a bit difficult... You imagine something like the following, I guess?
```
resource_list = [[s.strip() for s in c if isinstance(s, str) else s] for c in sheet.iter_rows(min_row=2, values_only=True)
                     if any(c) and any([re.search(r'[A-Za-z]+', elem) for elem in c if isinstance(elem, str)])]
```

`.strip()` must not be invoked on `None`, so I have to check if s is a string. But if s is not a string, I have to pass it to the method `_extract_row()` anyways, because it requires a fixed number of parameters.
Unfortunately, the above syntax doesn't work, so I have to do it the silly way.
And unfortunately, str(None) == 'None', which is a valid word that matches every regex that looks for words...

File: test/unittests/test_excel_to_properties.py (hunk 10 above)
Comment (jnussbaum) [reply test/unittests/test_excel_to_properties.py:35]: No, it can't get lost, because as long as at least one cell contains at least one letter, the entire row is processed.

File: knora/dsplib/utils/excel_to_json_resources.py (hunk 9 above)
Comment (irinaschubert) [reply knora/dsplib/utils/excel_to_json_resources.py:52]: I'm wondering if the whole thing would be easier if you import the data from an Excel into a pandas DataFrame. If I remember correctly, you added the dependency to pandas some time ago anyway, so you could use it for this part of the code, too? It would allow us to process the data more elegantly. This would involve a bigger refactoring of course.

File: test/unittests/test_excel_to_properties.py (hunk 10 above)
Comment (irinaschubert) [reply test/unittests/test_excel_to_properties.py:35]: It would be best to have one or several tests with values with these characters as well.

**Review by irinaschubert — CHANGES_REQUESTED** [review]
Could you add one or several more unit tests which test the special characters and numbers cases you discussed? Then we can be sure that it works as expected.

File: knora/dsplib/utils/excel_to_json_resources.py (hunk 9 above)
Comment (BalduinLandolt) [reply knora/dsplib/utils/excel_to_json_resources.py:52]: would the syntax work if you write something ternary like `[[s.strip() if isinstance(s, str) else s for s in c] ...`?  
But that's just fooling around with python at this stage :)

Pragmatically I'd say: Yes, pandas could be nice for this. And otherwise, maybe refactor the loading and cleaning of the data into one or two separate methods. Then you can also be a bit more explicit/lengthy about the whole process, so you don't need a one-liner that is so long that it doesn't fit on one line anymore. (Might be a good usecase for a Generator.)

File: knora/dsplib/utils/excel_to_json_properties.py [hunk 11]
```diff
@@ -1,12 +1,15 @@
 import json
 import os
+import re
 from typing import Any
 
 import jsonschema
-from openpyxl import load_workbook
+import pandas as pd
 
+from knora.dsplib.utils.excel_to_json_resources import _prepare_dataframe
```
Comment (BalduinLandolt) [inline knora/dsplib/utils/excel_to_json_properties.py:9]: since you're importing this here, it probably shouldn't be private (rule of thumb: everything that's used only in the same file should be private; everything used from a different file should not)

File: knora/dsplib/utils/excel_to_json_properties.py [hunk 12]
```diff
@@ -30,75 +33,110 @@ def validate_properties_with_schema(json_file: str) -> bool:
...
+    if row.get("en"):
+        labels["en"] = row["en"]
+    if row.get("de"):
+        labels["de"] = row["de"]
+    if row.get("fr"):
+        labels["fr"] = row["fr"]
+    if row.get("it"):
+        labels["it"] = row["it"]
+    if row.get("rm"):
+        labels["rm"] = row["rm"]
```
Comment (BalduinLandolt) [inline knora/dsplib/utils/excel_to_json_properties.py:63]: These could be nicely simplified. (same for the following, I just type out this one)

```suggestion
    languages = {"en", "de", "fr", "it", "rm"}
    labels = {lang: row[lang] for lang in languages if lang in row}
```

File: knora/dsplib/utils/excel_to_json_properties.py [hunk 13]
```diff
@@ -30,75 +33,110 @@ def validate_properties_with_schema(json_file: str) -> bool:
...
+    df = _prepare_dataframe(
+        df=df,
+        required_columns=["name", "super", "object", "gui_element"],
+        location_of_sheet=f"File '{excelfile}'")
+
+    # transform every row into a property
+    props: list[dict[str, Any]] = list()
+    for i, row in df.iterrows():
+        prop = _row2prop(row, i, excelfile)
+        props.append(prop)
```
Comment (BalduinLandolt) [inline knora/dsplib/utils/excel_to_json_properties.py:133]: list comprehension? :)

File: knora/dsplib/utils/excel_to_json_properties.py [hunk 14]
```diff
@@ -30,75 +33,110 @@ def validate_properties_with_schema(json_file: str) -> bool:
...
+
+    # write final list to JSON file if list passed validation
+    if _validate_properties_with_schema(json.loads(json.dumps(props, indent=4))):
+        with open(file=outfile, mode='w+', encoding='utf-8') as file:
+            file.write('"properties": ')
+            json.dump(props, file, indent=4)
```
Comment (BalduinLandolt) [inline knora/dsplib/utils/excel_to_json_properties.py:139]: is there actually a reason, why this turned into "pseudo-JSON" by adding `"properties": ` in front of it?

File: knora/dsplib/utils/excel_to_json_resources.py [hunk 15]
```diff
@@ -16,101 +16,151 @@ def validate_resources_with_schema(json_file: str) -> bool:
...
-    wb = load_workbook(excelfile, read_only=True)
-
-    # get overview
-    sheet = wb['classes']
-    resource_list = [c for c in sheet.iter_rows(min_row=2, values_only=True)]
+    # strip column headers and transform to lowercase, so that the script doesn't break when the headers vary a bit
+    new_df = df.rename(columns=lambda x: x.strip().lower())
+    required_columns = [x.strip().lower() for x in required_columns]
+    # strip every cell, and insert "" if there is no valid word in it
+    new_df = new_df.applymap(lambda x: str(x).strip() if pd.notna(x) and re.search(r"[\wäöü]", str(x)) else "")
```
[the remark this answers is not in the document]
Comment (BalduinLandolt) [reply knora/dsplib/utils/excel_to_json_resources.py:52]: also, since you added the german Umlaute, it would only be consequent to also add special characters from all other supported languages

File: knora/dsplib/utils/excel_to_json_resources.py [hunk 16]
```diff
@@ -16,101 +16,151 @@ def validate_resources_with_schema(json_file: str) -> bool:
...
+        new_df = new_df[pd.notna(new_df[req])]
+        new_df = new_df[[bool(re.search(r"[\wäöü]", x)) for x in new_df[req]]]
+    if len(new_df) < 1:
+        raise ValueError(f"{location_of_sheet} requires at least one row")
+    return new_df
+
+
+def _row2resource(row: pd.Series, excelfile: str) -> dict[str, Any]:
+    """
+    Method that takes a row from a pandas DataFrame, reads its content, and returns a dict object of the resource
```
Comment (BalduinLandolt) [inline knora/dsplib/utils/excel_to_json_resources.py:66]: I don't think one has to write "This method" in docstrings... I do it a lot too, but I think it's more a development-perspective than a usage-perspective. And in many methods it's not done, so it would be more consistent to leave it out here too

**Review by BalduinLandolt — APPROVED** [review]
I didn't look closely at the tests. Everything else seems good to me - all my comments are of cosmetic nature.

File: knora/dsplib/utils/excel_to_json_properties.py (hunk 14 above)
Comment (jnussbaum) [reply knora/dsplib/utils/excel_to_json_properties.py:139]: I think it's to be user friendly. This helps the user to better understand where exactly in the JSON s/he has to add the output. But besides, there is no real reason. It could theoretically be changed.

Comment (BalduinLandolt) [reply knora/dsplib/utils/excel_to_json_properties.py:115]: fair enough. I don't think it matters much

**Review by irinaschubert — APPROVED** [review]

File: test/e2e/test_tools.py [hunk 17]
```diff
@@ -155,14 +155,6 @@ def test_excel_to_json_list(self) -> None:
...
                         outfile='testdata/tmp/_lists-out.json')
 
-    def test_excel_to_json_resources(self) -> None:
-        resources_excel2json(excelfile='testdata/Resources.xlsx',
-                             outfile='testdata/tmp/_resources-out.json')
-
-    def test_excel_to_json_properties(self) -> None:
-        properties_excel2json(excelfile='testdata/Properties.xlsx',
-                              outfile='testdata/tmp/_properties-out.json')
-
```
[the remark this answers is not in the document]
Comment (jnussbaum) [reply test/e2e/test_tools.py]: Good point. I restored them, but deleted `test_validate_ontology()`, because it is not a dsp-tools command, but already contained in the `create` command
